sunxi-boot0.c

```diff
@@ -10,218 +11,640 @@
 #include <stdbool.h>
 #include <stdlib.h>
 #include <string.h>
+#include <errno.h>
+#include <stddef.h>
 
 #include "sunxi-fw.h"
 
-#define CHECKSUM_SEED	0x5f0a6c39
-
-#define BOOT0_CHECKSUM	3
-#define BOOT0_LENGTH	4
-#define BOOT0_DRAM_OFS	14
-#define OFS(x) ((x) + BOOT0_DRAM_OFS)
-
-enum dram_para {
-	DRAM_PARAM_NOT_USED = 0,
-	DRAM_CLK,
-	DRAM_TYPE,
-	DRAM_ZQ,
-	DRAM_ODT_EN,
-	DRAM_DX_ODT,
-	DRAM_DX_DRI,
-	DRAM_CA_DRI,
-	DRAM_PARA1, DRAM_PARA2,
-	DRAM_MR0, DRAM_MR1, DRAM_MR2, DRAM_MR3,
-	DRAM_MR4, DRAM_MR5, DRAM_MR6, DRAM_MR7,
-	DRAM_MR8, DRAM_MR9, DRAM_MR10, DRAM_MR11,
-	DRAM_MR12, DRAM_MR13, DRAM_MR14, DRAM_MR15,
-	DRAM_MR16, DRAM_MR17, DRAM_MR18, DRAM_MR19,
-	DRAM_MR20, DRAM_MR21, DRAM_MR22,
-	DRAM_TPR0, DRAM_TPR1, DRAM_TPR2, DRAM_TPR3,
-	DRAM_TPR4, DRAM_TPR5, DRAM_TPR6, DRAM_TPR7,
-	DRAM_TPR8, DRAM_TPR9, DRAM_TPR10, DRAM_TPR11,
-	DRAM_TPR12, DRAM_TPR13,
-	NR_DRAM_PARAMS
+#define __maybe_unused  __attribute__((unused))
+
+#define EGON_CHECKSUM_SEED	0x5f0a6c39
+
+struct egon_header {
```

Comment:
Because it is a struct in the allwinner code, and a struct vastly improves readability.

Indexes into arrays are a total pain, and impossible to read.

Reply:
The egon header, and the secondary header are definitely not hw registers, and are (mostly) actual structs in vendor code.